chasm/visibility_manager.go

```diff
@@ -0,0 +1,156 @@
+//go:generate mockgen -package $GOPACKAGE -source $GOFILE -destination visibility_manager_mock.go
+
+package chasm
+
+import (
+	"context"
+	"reflect"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/common/payload"
+	"google.golang.org/protobuf/proto"
+)
+
+type VisibilityManager interface {
+	ListExecutions(
+		context.Context,
+		reflect.Type,
+		*ListExecutionsRequest,
+	) (*ListExecutionsResponse[*commonpb.Payload], error)
+
+	CountExecutions(
+		context.Context,
+		reflect.Type,
+		*CountExecutionsRequest,
+	) (*CountExecutionsResponse, error)
+}
+
+type ExecutionInfo[M proto.Message] struct {
+	BusinessID             string
+	RunID                  string
+	StartTime              time.Time
+	CloseTime              time.Time
+	HistoryLength          int64
+	HistorySizeBytes       int64
+	StateTransitionCount   int64
+	ChasmSearchAttributes  SearchAttributesMap
+	CustomSearchAttributes map[string]*commonpb.Payload
+	Memo                   *commonpb.Memo
+	ChasmMemo              M
+}
+
+type ListExecutionsRequest struct {
+	NamespaceID   string
+	NamespaceName string
+	Query         string
+	PageSize      int
+	NextPageToken []byte
+}
+
+type ListExecutionsResponse[M proto.Message] struct {
+	Executions    []*ExecutionInfo[M]
+	NextPageToken []byte
+}
+
+type CountExecutionsRequest struct {
+	NamespaceID   string
+	NamespaceName string
+	Query         string
+}
+
+type CountExecutionsResponse struct {
+	Count int64
+}
+
+// ListExecutions lists the executions of a CHASM archetype given an initial query.
+// The query string can specify any combination of CHASM, custom, and predefined/system search attributes.
+// The generic parameter C is the CHASM component type used for executions and search attribute filtering.
+// The generic parameter M is the type of the memo payload to be unmarshaled from the execution.
+// PageSize is required, must be greater than 0.
+// NextPageToken is optional, set on subsequent requests to continue listing the next page of executions.
+// Note: For CHASM executions, TemporalNamespaceDivision is the predefined search attribute
+// that is used to identify the archetype of the execution.
+// If the query string does not specify TemporalNamespaceDivision, the archetype C of the request will be used to filter the executions.
+// If the initial query already specifies TemporalNamespaceDivision, the archetype C of the request will
+// only be used to get the registered SearchAttributes.
+func ListExecutions[C Component, M proto.Message](
+	ctx context.Context,
+	request *ListExecutionsRequest,
```

Comment:
I missed this in the initial review, can we make this not a pointer? It's slightly nicer/safer to know that the arguments are copied when the function is called. Same for the response and count BTW.